### DISCUSSION TITLE
How to save the signin state

### DISCUSSION THREAD

**Discussion by alex-user-go** [xelaj/mtproto#67] [Q&A] [answered]
Hello! I need some help in saving signin state. I could not figure out how not to auth every time I start my app. Is it possible? Could you provide me with a code example?

**Comment by quenbyako** [comment, marked as answer]
Hey @alex-user-go! Telegram config has `SessionFile` parameter, in this path mtproto client saves your session. telegram authorization is too unintuitive but here is the simplest explaination: when you connect to tg servers, mtproto client must store specific session credentials, so when you send `auth.signIn` request, you authorize from _specific session_, not in your app. (yeah, pretty strange but it is). after your session was authorized, you can send other requests from this session. Hope that i explained you as simple as i can.

Also, telegram doesn't like this way, when you signin'ing many times per short duration, telegram block you for 24+ hours due to spamming. soooo, probably you don't need to authorize this way. I agree that authorization system in tg api is not perfect, but we have what we have.

**Reply by alex-user-go** [reply]
Thank you for such a quick answer, but still I can't figure it out. After signing in (`AuthSignIn`) evertything works fine. Then I restart my app, call `telegram.NewClient` (the same sessionFile) and call, say, `AccountInitTakeoutSession` I have _The key is not registered in the system (code 401)_. Also I don't get the meaning of this line https://github.com/xelaj/mtproto/blob/c025fd4ddb069118e638002b31d4ec94c6de6320/examples/auth/main.go#L49 as `telegram.NewClient` always creates a sessionFile (if it didn't exist). Am I right?

**Reply by quenbyako** [reply]
this is super stupid bug, i'll fix it right now, lol :rofl:

**Reply by alex-user-go** [reply]
And now it will never reach `AuthSignIn`, cause after `telegram.NewClient` there is always a sessionFile, right?

**Reply by quenbyako** [reply]
when you have `session.json` file (see how `utils.PrepareAppStorageForExamples()` works), then yes, mtproto object will see that session is exist and will not recreate new session.

**Reply by alex-user-go** [reply]
Any ideas about error 401? 

> 
> 
> Thank you for such a quick answer, but still I can't figure it out. After signing in (`AuthSignIn`) evertything works fine. Then I restart my app, call `telegram.NewClient` (the same sessionFile) and call, say, `AccountInitTakeoutSession` I have _The key is not registered in the system (code 401)_. Also I don't get the meaning of this line
> 
> https://github.com/xelaj/mtproto/blob/c025fd4ddb069118e638002b31d4ec94c6de6320/examples/auth/main.go#L49
> as `telegram.NewClient` always creates a sessionFile (if it didn't exist). Am I right?

**Comment by alex-user-go** [comment]
If anyone is struggling with this - it turned out that the testhost requires registration even though your phone is registered with a serverhost.

**Reply by quenbyako** [reply]
no way... 😱😱😱
